File: exercises/day18-Risky-Sleigh/scala/src/main/scala/core/control/BestMagicalPerformancePowerUnitFactory.scala
package core.control

import external._

class BestMagicalPerformancePowerUnitFactory(
                                              private val allReindeers: List[Reindeer],
                                              private val availableAmplifierByMagicalPower: Map[Int, AmplifierType]
                                            ) extends PowerUnitFactory {

  override def bringAllReindeers(): List[ReindeerPowerUnit] = {
    val allReindeerByMagicalPower = allReindeers
      .sortBy(_.getMagicPower)(Ordering[Int].reverse)

    allReindeerByMagicalPower
      .zipWithIndex
      .map { case (reindeer, index) =>
        attachPowerUnit(reindeer, index + 1)
      }
  }

  private def attachPowerUnit(reindeer: Reindeer, indexOfMagicalPower: Int): ReindeerPowerUnit = {
    generatePowerUnit(
      reindeer,
      availableAmplifierByMagicalPower.getOrElse(indexOfMagicalPower, AmplifierType.Basic)
    )
  }

  private def generatePowerUnit(
                                 reindeer: Reindeer,
                                 amplifierToAttach: AmplifierType
                               ): ReindeerPowerUnit = {
    new ReindeerPowerUnit(reindeer, new MagicPowerAmplifier(amplifierToAttach))
  }
}

File: exercises/day18-Risky-Sleigh/kotlin/src/main/kotlin/core/control/BestMagicalPerformancePowerUnitFactory.kt
package core.control

import external.deer.Reindeer

class BestMagicalPerformancePowerUnitFactory(
    private val allReindeers: List<Reindeer>,
    private val availableAmplifierByMagicalPower: Map<Int, AmplifierType>
) : PowerUnitFactory {

    override fun bringAllReindeers(): List<ReindeerPowerUnit> {
        val allReindeerByMagicalPower = allReindeers
            .sortedByDescending { it.magicPower }

        return allReindeerByMagicalPower
            .mapIndexed { index, reindeer ->
                attachPowerUnit(reindeer, index + 1)
            }
    }

    private fun attachPowerUnit(reindeer: Reindeer, indexOfMagicalPower: Int): ReindeerPowerUnit {
        return generatePowerUnit(
            reindeer,
            availableAmplifierByMagicalPower.getOrDefault(indexOfMagicalPower, AmplifierType.BASIC)
        )
    }

    private fun generatePowerUnit(
        reindeer: Reindeer,
        amplifierToAttach: AmplifierType
    ): ReindeerPowerUnit {
        return ReindeerPowerUnit(reindeer, MagicPowerAmplifier(amplifierToAttach))
    }
}